torchao/quantization/quant_primitives.py

```diff
@@ -69,6 +72,41 @@ class ZeroPointDomain(Enum):
     })
 
 
+# def register_custom_op(name: str):
+#     from torch._inductor.decomposition import register_decomposition
+
+#     def decorator(fn):
+#         if TORCH_VERSION_AFTER_2_5:
+#             opdef = torch.library.custom_op(name, mutates_args=())(fn)
+#             opdef.register_fake(fn)
+#             register_decomposition([opdef._opoverload])(fn)
+#             return opdef
+#         else:
+#             return fn
+
+#     return decorator
+
+quant_lib = torch.library.Library("quant", "FRAGMENT")
+
+def register_custom_op(lib, schema: str):
+    from torch._inductor.decomposition import register_decomposition
+
+    def decorator(fn):
+        if TORCH_VERSION_AFTER_2_5:
+            # TODO: change order
+            lib_namespace = lib.ns
+            op_name = schema.split("(")[0]
+            lib.define(schema)
+            lib.impl(op_name, fn, "CompositeImplicitAutograd")
+            op = getattr(getattr(torch.ops, lib_namespace), op_name)
+            register_decomposition([op])(fn)
+            return fn
```

Comment:
This needs to `return op` instead of `return fn` in order to replace the function with the custom op. Unfortunately that causes stuff to error out with:
```
  File "/home/rzou/dev/opt-cuda4/ao/torchao/quantization/subclass.py", line 661, in __torch_function__
    return torch.nn.functional.linear(aqt, original_weight_tensor, bias)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzou/dev/opt-cuda4/ao/torchao/dtypes/affine_quantized_tensor.py", line 241, in __torch_function__
    return _ATEN_OP_OR_TORCH_FN_TABLE[cls][func](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzou/dev/opt-cuda4/ao/torchao/dtypes/utils.py", line 25, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzou/dev/opt-cuda4/ao/torchao/dtypes/affine_quantized_tensor.py", line 685, in functional_linear
    input_tensor = input_tensor.dequantize()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzou/dev/opt-cuda4/ao/torchao/dtypes/affine_quantized_tensor.py", line 161, in dequantize
    return dequantize_affine(int_data, self.block_size, scale, zero_point, int_data.dtype, self.quant_min, self.quant_max,
 self.zero_point_domain, output_dtype=output_dtype)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rzou/dev/opt-cuda4/ao/torchao/quantization/quant_primitives.py", line 304, in dequantize_affine
    return _dequantize_affine(
           ^^^^^^^^^^^^^^^^^^^
  File "/home/rzou/dev/opt-cuda4/env/lib/python3.11/site-packages/torch/_ops.py", line 1064, in __call__
    return self_._op(*args, **(kwargs or {}))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: _dequantize_affine() takes from 5 to 8 positional arguments but 9 were given

----------------------------------------------------------------------
Ran 1 test in 0.068s
```
thoughts?

Reply:
is it OK for the decorator to return `op` instead of `fn`?

Reply:
yes